exchange/utils.go

```diff
@@ -902,6 +919,60 @@ func mergeBidderRequests(allBidderRequests []BidderRequest, bidderNameToBidderRe
 	return allBidderRequests
 }
 
+func setLegacyGDPRFromGPP(r *openrtb2.BidRequest, gpp gpplib.GppContainer) {
+	if r.Regs != nil && r.Regs.GDPR == nil {
+		if r.Regs.GPPSID != nil {
+			// Set to 0 unless SID exists
+			regs := *r.Regs
+			regs.GDPR = ptrutil.ToPtr[int8](0)
+			for _, id := range r.Regs.GPPSID {
+				if id == int8(gppConstants.SectionTCFEU2) {
+					regs.GDPR = ptrutil.ToPtr[int8](1)
+				}
+			}
+			r.Regs = &regs
+		}
+	}
+
+	if r.User == nil || len(r.User.Consent) == 0 {
+		for i, id := range gpp.SectionTypes {
+			if id == gppConstants.SectionTCFEU2 {
+				if r.User == nil {
+					r.User = &openrtb2.User{
+						Consent: gpp.Sections[i].GetValue(),
+					}
+				} else {
+					user := *r.User
+					user.Consent = gpp.Sections[i].GetValue()
+					r.User = &user
+				}
+			}
+		}
+	}
+
+}
+func setLegacyUSPFromGPP(r *openrtb2.BidRequest, gpp gpplib.GppContainer) {
+	if r.Regs == nil {
+		return
+	}
+
+	if len(r.Regs.USPrivacy) > 0 || r.Regs.GPPSID == nil {
+		return
+	}
+	for _, id := range r.Regs.GPPSID {
+		if id == int8(gppConstants.SectionUSPV1) {
+			for i, id := range gpp.SectionTypes {
+				if id == gppConstants.SectionUSPV1 {
```

Comment:
This check duplicates L963: `if id == int8(gppConstants.SectionUSPV1)`

Reply:
It should not be a duplicate, but it is not clear that the second `id` is a new one in the inner scope. I will rename to ensure Go treats it correctly and avoids this confusion for readers.

Reply:
This also has the assumption that the two arrays line up from before, so clearing that up as well.

Reply:
Thank you, looks good